src/mpid/ch4/shm/ipc/src/ipc_send.h

```diff
@@ -13,6 +13,42 @@
 #include "../gpu/gpu_post.h"
 #include "ipc_p2p.h"
 
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+cvars:
+    - name        : MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR
+      category    : CH4
+      type        : boolean
+      default     : true
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        If an address is used more than once in the last ten send operations,
+        map it for IPC use even if it is below the IPC threshold.
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX bool MPIDI_IPCI_is_repeat_addr(void *addr)
+{
+    if (!MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR) {
+        return false;
+    }
+
+    static void *repeat_addr[10] = { 0 };
+    static int addr_idx = 0;
+
+    for (int i = 0; i < 10; i++) {
+        if (addr == repeat_addr[i]) {
+            return true;
+        }
+    }
+
+    repeat_addr[addr_idx] = addr;
+    addr_idx = (addr_idx + 1) % 10;
+    return false;
+}
```

Comment:
This should help identify repeat addresses being used. So where should we ultimately set the bounds? Set IPC threshold even higher (e.g. 1MB) and use the repeat check for anything smaller than that?

Reply:
Let's set the IPC thresh 1MB and check OSU latency. We should the hit IPC across all sizes now since OSU always repeat the buffer address.

Reply:
First run is with the threshold set to 1MB + repeat address check. Second run is with the same threshold and repeat check disabled. You can see IPC kick in at 1MB.
```
[raffenet@pmrs-gpu-240-01]~/osu-micro-benchmarks-6.1/c/mpi/pt2pt% ~/proj/mpich-gpu-shm/i/bin/mpiexec -l -n 2 -bind-to package ./osu_latency D D |& tee out.txt
[0] # OSU MPI-CUDA Latency Test v6.1
[0] # Send Buffer on DEVICE (D) and Receive Buffer on DEVICE (D)
[0] # Size          Latency (us)
[0] 0                       3.20
[0] 1                      17.46
[0] 2                      14.29
[0] 4                      12.36
[0] 8                      12.39
[0] 16                     12.37
[0] 32                     12.36
[0] 64                     12.37
[0] 128                    12.38
[0] 256                    12.38
[0] 512                    12.39
[0] 1024                   12.40
[0] 2048                   12.61
[0] 4096                   12.63
[0] 8192                   13.12
[0] 16384                  14.10
[0] 32768                  15.43
[0] 65536                  18.63
[0] 131072                 25.09
[0] 262144                 37.85
[0] 524288                 63.41
[0] 1048576               114.54
[0] 2097152               217.37
[0] 4194304               420.69
[raffenet@pmrs-gpu-240-01]~/osu-micro-benchmarks-6.1/c/mpi/pt2pt% export MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR=0
[raffenet@pmrs-gpu-240-01]~/osu-micro-benchmarks-6.1/c/mpi/pt2pt% ~/proj/mpich-gpu-shm/i/bin/mpiexec -l -n 2 -bind-to package ./osu_latency D D |& tee out.txt
[0] # OSU MPI-CUDA Latency Test v6.1
[0] # Send Buffer on DEVICE (D) and Receive Buffer on DEVICE (D)
[0] # Size          Latency (us)
[0] 0                       3.28
[0] 1                      31.65
[0] 2                      26.90
[0] 4                      22.22
[0] 8                      22.17
[0] 16                     22.19
[0] 32                     22.18
[0] 64                     22.22
[0] 128                    22.29
[0] 256                    22.36
[0] 512                    22.47
[0] 1024                   22.82
[0] 2048                   23.37
[0] 4096                   24.41
[0] 8192                   26.50
[0] 16384                  46.37
[0] 32768                  62.64
[0] 65536                  95.52
[0] 131072                160.99
[0] 262144                293.19
[0] 524288                555.96
[0] 1048576               114.08
[0] 2097152               216.30
[0] 4194304               420.14
```

Reply:
Nice!

Reply:
1MB without IPC latency will be around 1 ms, which matches the order of IPC registration overhead. I think it is a good threshold.